huggingface_pipelines/pipeline.py

```diff
@@ -0,0 +1,79 @@
+from abc import ABC, abstractmethod
+from typing import Dict, Any
+import logging
+from .pipeline_config import PipelineConfig
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+class Pipeline(ABC):
+    """
+    Abstract base class for different pipelines.
+    """
+
+    def __init__(self, config: PipelineConfig):
+        self.config = config
+        self.results = []
+
```

Comment:
for abstract class I would implement:
```python
def __call__(self, dataset: HFDataset) -> dataset:
       .... 
```
in case of charding, you will do something like :
```python
def __call__(self, dataset: HFDataset) -> dataset:
     return dataset.shard(num_shards=self.config.num_shards, index=self.config.shard_id)
```
so in overall combination will look like :
```
dataset = ReadPipeline(read_config)(None) # with could contain sharding logic 
dataset = SonarEncoderPipeline(sonar_encoder_config)(dataset)
dataset = SonarDecoderPipeline(...)(dataset)
dataset = BleuScorePipeline(...)(dataset)
resulst = dataset.save_on_disk(...)
```
as main principle, you always add new columns and never remove the existing ones (incremental).